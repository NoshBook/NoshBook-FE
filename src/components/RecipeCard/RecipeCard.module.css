.container {
  display: flex;
  flex-direction: column;
  width: 350px;
  background-color: var(--light-black);
  padding: 20px;
  border-radius: 10px;
  min-height: 400px;
  box-shadow: 0px 5px 15px rgb(18, 17, 17);
  cursor: pointer;
  gap: 10px;
}

.cardimg {
  width: 100%;
  height: 200px;
  object-fit: cover;
}

.cardh2 {
  margin-bottom: 10px;
}

.cookbookbuttons {
  display: flex;
  justify-content: space-between;
  margin-top: 10px;
  margin-bottom: 10px;
}

.cookbookbuttons button {
  box-shadow: none;
}
