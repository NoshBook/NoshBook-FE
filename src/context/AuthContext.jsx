import { createContext, useContext, useMemo, useState, useEffect } from 'react';
import { getUser } from '../services/users';

const AuthContext = createContext();

const AuthProvider = ({ children }) => {
  const [user, setUser] = useState({ showUserContent: false });
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const getCurrentUser = async () => {
      const res = await getUser();
      console.log(res);
      // The nullish coalescing operator prevents user from being set to undefined
      setUser(res ?? { showUserContent: false });
      setLoading(false);
    };
    getCurrentUser();
  }, []);

  const updateUserPreference = () => {
    setUser((prevState) => {
      return {
        ...prevState,
        showUserContent: !prevState.showUserContent,
      };
    });
  };

  const value = useMemo(
    () => ({ user, setUser, updateUserPreference }),
    [user],
  );
  if (loading) {
    // This component will lose its state on page refreshes and manually entered URLs.
    // Adding this wait allows the user to be fetched before rendering any child
    // PrivateRoutes, which prevents logged in users from being erroneously redirected.
    return (
      <div className="authcontextloading">
        <h2>Loading...</h2>
      </div>
    );
  } else {
    return (
      <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
    );
  }
};

const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth not used within Auth Provider');
  }
  return context;
};

export { AuthContext, AuthProvider, useAuth };
