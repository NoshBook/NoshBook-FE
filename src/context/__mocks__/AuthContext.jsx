import { createContext, useContext, useMemo, useState, useEffect } from 'react';
import { getUser } from '../../services/users.js';

const AuthContext = createContext();

const AuthProvider = ({ mockUser = { showUserContent: false }, children }) => {
  console.log('<<<<<<AUTHPROVIDER MOCK INITIATED>>>>>>');
  const [user, setUser] = useState({ ...mockUser });
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const getCurrentUser = async () => {
      const res = await getUser();
      console.log('MOCK USER: ', res);
      setUser(res);
      setLoading(false);
    };
    getCurrentUser();
  }, []);

  const updateUserPreference = () => {
    setUser((prevState) => {
      return {
        ...prevState,
        showUserContent: !prevState.showUserContent,
      };
    });
  };

  const value = useMemo(
    () => ({ user, setUser, updateUserPreference }),
    [user],
  );
  if (loading) {
    return <h2>Loading...</h2>;
  } else {
    return (
      <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
    );
  }
};

const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth not used within Auth Provider');
  }
  return context;
};

export { AuthContext, AuthProvider, useAuth };
