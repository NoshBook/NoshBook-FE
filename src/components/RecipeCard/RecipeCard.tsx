import React from 'react';
import { useAuth } from '../../context/AuthContext';
import { BrowseRecipe } from '../../views/Browse/interfaces/BrowseRecipe';
import { Rating } from 'react-simple-star-rating';
import { useState } from 'react';
import DaysMenu from '../DaysMenu/DaysMenu';
import { useNavigate } from 'react-router-dom';
import styles from './RecipeCard.module.css';
import { motion } from 'framer-motion';

import {
  AiOutlinePlusCircle,
  AiOutlineEdit,
  AiOutlineMinusCircle,
} from 'react-icons/ai';

interface RecipeCardProps {
  recipe: BrowseRecipe;
  isCookbookView?: boolean;
  plannerToggle?: boolean;
  setPlannerToggle?: React.Dispatch<React.SetStateAction<boolean>>;
  handleRemoveFromCookbookClick?: (id: string) => void;
  handleAddToPlannerClick?: (day: string, recipeId: string) => void;
}

export default function RecipeCard({
  recipe,
  isCookbookView,
  handleRemoveFromCookbookClick,
  handleAddToPlannerClick,
}: RecipeCardProps) {
  const [plannerToggle, setPlannerToggle] = useState(false);
  const navigate = useNavigate();
  const { user } = useAuth();
  const { name, rating, image, description } = recipe;
  return (
    <article className={styles.container}>
      <h2 className={styles.cardh2}>{name}</h2>

      <Rating
        initialValue={rating * 20}
        ratingValue={rating * 20}
        size={25}
        readonly
      />

      {isCookbookView && (
        <section aria-label="Recipe Options" className={styles.cookbookbuttons}>
          <motion.button
            whileHover={{ scale: 1.02 }}
            aria-label="Remove recipe from cookbook."
            onClick={(e) => {
              e.stopPropagation();
              handleRemoveFromCookbookClick?.(recipe.id);
            }}
          >
            <AiOutlineMinusCircle color={'var(--orange)'} />
            Cookbook
          </motion.button>
          <div className={styles.daysmenucontainer}>
            <motion.button
              whileHover={{ scale: 1.02 }}
              aria-label="Toggles planner options display."
              aria-pressed={plannerToggle}
              onClick={(e) => {
                e.stopPropagation();
                setPlannerToggle((prevState) => !prevState);
              }}
              disabled={user.id ? false : true}
              // --- Can be removed
              title={
                user.id
                  ? 'Click to interact with recipe options'
                  : 'Login to interact with recipe options'
              }
              // ---
            >
              <AiOutlinePlusCircle color={'var(--blue)'} /> Planner
            </motion.button>

            {plannerToggle && (
              <DaysMenu
                handleAddToPlanner={handleAddToPlannerClick}
                recipeId={recipe.id}
                setPlannerToggle={setPlannerToggle}
              />
            )}
          </div>

          <motion.button
            whileHover={{ scale: 1.02 }}
            aria-label="Edit Recipe. Redirects to new page."
            onClick={(e) => {
              e.stopPropagation();
              // may need to be updated after create/edit is pushed
              navigate(`/recipes/edit/${recipe.id}`);
            }}
          >
            <AiOutlineEdit color={'var(--blue)'} /> Edit
          </motion.button>
        </section>
      )}

      <img className={styles.cardimg} src={image} alt={name} />

      <p
        className={
          description?.length < 200
            ? styles.cardp
            : styles.cardpsmalldescription
        }
      >
        {description}
      </p>
    </article>
  );
}
