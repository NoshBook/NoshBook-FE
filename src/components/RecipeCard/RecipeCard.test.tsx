import RecipeCard from './RecipeCard';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { AuthProvider } from '../../context/AuthContext';

jest.mock('../../context/AuthContext');

const mockRecipe = {
  description: 'mock recipe description.',
  id: '1',
  image: 'https://placekitten.com/100/100',
  ingredients: ['mock-recipe-ingredient'],
  instructions: ['mock-recipe-instruction'],
  name: 'mock-recipe-name',
  rating: 5,
  servings: '5',
  tags: ['mock-recipe-tag'],
  totalTime: 'mock-recipe-time',
  ownerId: '1',
};

describe('recipeCard component snapshots', () => {
  it('renders a browse-all view recipe card', async () => {
    const { container } = render(
      <AuthProvider>
        <MemoryRouter>
          <RecipeCard recipe={mockRecipe} />
        </MemoryRouter>
      </AuthProvider>,
    );

    await screen.findByText('mock-recipe-name');

    expect(container).toMatchSnapshot();
  });

  it('renders a cookbook view recipe card', async () => {
    const { container } = render(
      <AuthProvider>
        <MemoryRouter>
          <RecipeCard
            recipe={mockRecipe}
            handleAddToPlannerClick={jest.fn()}
            handleRemoveFromCookbookClick={jest.fn()}
            isCookbookView={true}
          />
        </MemoryRouter>
      </AuthProvider>,
    );

    screen.getByText(/loading/i);
    await screen.findByText('mock-recipe-name');

    expect(container).toMatchSnapshot();
  });
});
