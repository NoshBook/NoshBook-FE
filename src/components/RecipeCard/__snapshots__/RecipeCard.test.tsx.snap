// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recipeCard component snapshots renders a browse-all view recipe card 1`] = `
<div>
  <article
    class="container"
  >
    <h2
      class="cardh2"
    >
      mock-recipe-name
    </h2>
    <span
      style="display: inline-block; direction: ltr;"
    >
      <span
        aria-hidden="true"
        class="react-simple-star-rating"
        style="position: relative; display: inline-block; overflow: hidden; white-space: nowrap; vertical-align: middle; user-select: none;"
      >
        <span
          class="empty-icons"
          style="display: inline-block; color: rgb(204, 204, 204);"
        >
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
        </span>
        <span
          class="filled-icons"
          style="position: absolute; top: 0px; left: 0px; color: rgb(255, 188, 11); overflow: hidden; white-space: nowrap; display: inline-block; width: 100%;"
          title="100 out of 5"
        >
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
        </span>
      </span>
    </span>
    <img
      alt="mock-recipe-name"
      class="cardimg"
      src="https://placekitten.com/100/100"
    />
    <p
      class="cardp"
    >
      mock recipe description.
    </p>
  </article>
</div>
`;

exports[`recipeCard component snapshots renders a cookbook view recipe card 1`] = `
<div>
  <article
    class="containercookbook"
  >
    <h2
      class="cardh2"
    >
      mock-recipe-name
    </h2>
    <span
      style="display: inline-block; direction: ltr;"
    >
      <span
        aria-hidden="true"
        class="react-simple-star-rating"
        style="position: relative; display: inline-block; overflow: hidden; white-space: nowrap; vertical-align: middle; user-select: none;"
      >
        <span
          class="empty-icons"
          style="display: inline-block; color: rgb(204, 204, 204);"
        >
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
        </span>
        <span
          class="filled-icons"
          style="position: absolute; top: 0px; left: 0px; color: rgb(255, 188, 11); overflow: hidden; white-space: nowrap; display: inline-block; width: 100%;"
          title="100 out of 5"
        >
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
          <svg
            class="star-svg"
            fill="currentColor"
            height="25"
            viewBox="0 0 24 24"
            width="25"
          >
            <path
              d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
              fill="currentColor"
              stroke="none"
              stroke-miterlimit="10"
              stroke-width="0"
            />
          </svg>
        </span>
      </span>
    </span>
    <section
      aria-label="Recipe Options"
      class="cookbookbuttons"
    >
      <button
        aria-label="Remove recipe from cookbook."
      >
        <svg
          color="var(--orange)"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M696 480H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h368c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
          />
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
          />
        </svg>
        Cookbook
      </button>
      <div
        class="daysmenucontainer"
      >
        <button
          aria-label="Toggles planner options display."
          aria-pressed="false"
          disabled=""
          title="Login to interact with recipe options"
        >
          <svg
            color="var(--blue)"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
            />
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
          </svg>
           Planner
        </button>
      </div>
      <button
        aria-label="Edit Recipe. Redirects to new page."
      >
        <svg
          color="var(--blue)"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 0 0 0-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 0 0 9.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
          />
        </svg>
         Edit
      </button>
    </section>
    <img
      alt="mock-recipe-name"
      class="cardimg"
      src="https://placekitten.com/100/100"
    />
    <p
      class="cardp"
    >
      mock recipe description.
    </p>
  </article>
</div>
`;
